File: lib/aoc24/day01.ex
defmodule Aoc24.Day01 do
  def part1(file) do
    file
    |> parse_input()
    |> Enum.map(&Enum.sort/1)
    |> then(fn [a, b] -> Enum.zip(a, b) end)
    |> Enum.map(fn {a, b} -> abs(a - b) end)
    |> Enum.sum()
  end

  def part2(file) do
    file
    |> parse_input()
    |> then(fn [left, right] ->
      right = Enum.frequencies(right)

      left
      |> Enum.map(&(&1 * Map.get(right, &1, 0)))
      |> Enum.sum()
    end)
  end

  defp parse_input(file) do
    file
    |> File.read!()
    |> String.split("\n")
    |> Enum.reject(&(&1 == ""))
    |> Enum.map(&String.split(&1, "   "))
    |> Enum.zip_with(& &1)
    |> Enum.map(fn nums -> Enum.map(nums, &String.to_integer/1) end)
  end
end

File: src/aoc24/day01.clj
(ns aoc24.day01)

(defn parse-input [file]
  (->> (re-seq #"\d+" (c/get-input file))
       (map parse-long)
       (partition 2)
       (apply map vector)))

(defn part1 [file]
  (->> (parse-input file)
       (map sort)
       ((fn [[a b]] (interleave a b)))
       (partition 2)
       (map (fn [[a b]] (- a b)))
       (map abs)
       (reduce +)))

(defn part2 [file]
  (let [[a b] (parse-input file)
        b (frequencies b)]
    (->> a
         (map #(* % (get b % 0)))
         (reduce +))))

